data-prepper-core/src/integrationTest/java/org/opensearch/dataprepper/plugins/InMemorySource.java

```diff
@@ -27,23 +32,39 @@ public class InMemorySource implements Source<Record<Event>> {
     private final InMemorySourceAccessor inMemorySourceAccessor;
     private boolean isStopped = false;
     private Thread runningThread;
+    private final AcknowledgementSetManager acknowledgementSetManager;
+    private boolean enabledAcks;
 
     @DataPrepperPluginConstructor
     public InMemorySource(
             final InMemoryConfig inMemoryConfig,
+            final EventFactory eventFactory,
+            final AcknowledgementSetManager acknowledgementSetManager,
             final InMemorySourceAccessor inMemorySourceAccessor) {
         testingKey = inMemoryConfig.getTestingKey();
 
         this.inMemorySourceAccessor = inMemorySourceAccessor;
+        this.inMemorySourceAccessor.setEventFactory(eventFactory);
+        this.acknowledgementSetManager = acknowledgementSetManager;
+        this.enabledAcks = inMemoryConfig.getEndToEndAcknowledgements();
     }
 
     @Override
     public void start(final Buffer<Record<Event>> buffer) {
-        runningThread = new Thread(new SourceRunner(buffer));
+        if (enabledAcks) {
```

Comment:
I think it might be advantageous to make a new class which supports acknowledgements - perhaps `InMemoryAcknowledgementSource`.

You could do some refactoring to have both `InMemorySource` and `InMemoryAcknowledgementSource` inherit from a common base class.

It could simplify the code for both of these and make them easier to manage.

Reply:
That's how I implemented initially but I think there is lot of duplicate code, so I merged it. Also, we want the plugin to be same but behave differently when `end_to_end_acknowledgements:true` is specified to the same source.

Perhaps, I should have two different `Runnables` like `SourceRunner` and `AcknowledgementSourceRunner`

Reply:
Nice!